### PR TITLE
Week 4 cathrine carreon

### PR DESCRIPTION
**Pull request by CathrineC**
# Submitting Week 4 Assessment

### ✅ Completed Files
Mark the following as confirmation that the files in this repository are complete.
- [x] JavaScript Coding Challenges
- [x] Ruby Coding Challenges
- [x] Interview Questions

### 📝 Feedback
If you would like specific feedback on a particular problem please leave details here:

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -21,9 +21,45 @@ const colors1 = ["purple", "blue", "green", "yellow", "pink"]
 const colors2 = ["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saffron"]
 // Expected output example (can be a different order): ["saffron", "aquamarine", "periwinkle", "indigo", "ochre"]
 
+describe("shuffleWords", () => {
+    it("returns the remaining content of an array shuffled", () => {
+        expect(colors1.length-1).toEqual(shuffleWords(colors1).length)
```
Comment (SunkissedQueen) [inline code-challenges.test.js:26]: Creative approach. Testing that the length has changed. I would say that is a great additional element to test. The necessary expect statement to truly test your function is the one that includes the function call. 
I like that you attempted the not arrayContaining matcher. The structure here is conveying that you are not expecting the array to contain that value that was removed. 
Just missing the `expect.arrayContaining`. 
Here's an example
`expect(shuffleWords(colors1)).not.toEqual(expect.arrayContaining["purple"])`

File: code-challenges.test.js [hunk 2]
```diff
@@ -21,9 +21,45 @@ const colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
+// - Expected Output: Return the remaining elements in the array shuffled.
+
+const shuffleWords = (colors) => {
+    colors.reverse().pop()
+    colors.sort(function(){
+        return Math.random() - 0.5
+    }) 
+    return colors
+}
+
```
Comment (SunkissedQueen) [inline code-challenges.test.js:59]: great work. I suggest going back for a refactor

File: code-challenges.test.js [hunk 3]
```diff
@@ -34,9 +70,34 @@ const votes1 = {upVotes: 13, downVotes: 2}
...
+// - Input: object with numbers as values
+// - gather the values in an array
+// - deduct the value on index 1 from the value on index 0
+// - Return the net total of votes
+
+const totalVotes = (number) => {
+    let votes = Object.values(number)
+    return votes[0] - votes[1]
+}
+            // PASS  ./code-challenges.test.js
```
Comment (SunkissedQueen) [inline code-challenges.test.js:98]: love this approach

File: code-challenges.test.js [hunk 4]
```diff
@@ -46,5 +107,50 @@ const dataArray1 = ["array", "object", "number", "string", "Boolean"]
...
+const noDoubleValues = (arrOne, arrTwo) => {
+    let twoArr = [...arrOne, ...arrTwo];
+    let combined = twoArr.filter((data, dupData) => 
+    twoArr.indexOf(data) === dupData)
+    return combined;
+}
+            // PASS  ./code-challenges.test.js
+            // noDoubleValues
+            //   ✓ returns one array with no duplicate values (2 ms)
+
```
Comment (SunkissedQueen) [inline code-challenges.test.js:156]: Cathrine, you are doing great work here. If you have time, I would recommend going back through your functions to refactor your code using tools that will take less lines.
For instance, search "javascript how to remove duplicate values from arrays"

File: interview-questions.md [hunk 5]
```diff
@@ -8,42 +8,78 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-Researched answer:
+Researched answer: 
+Based on what I read from resources online, Functional programming is based on pure function. It takes an input, applies a process and returns a certain output or value. Unlike, Object-Oriented Programming, Functional programming doesn't use state because it functions in its own scope. It works independetly and avoids mutable data. This way, it can avoid what they say "side effects" but can take inputs and still reusable. While Functional Programming is based on pure functions, Object-Oriented Programming is based on objects and classes. OOP code can be encapsulated depending on how you set it. JavaScript makes use of both OOP and Functional Programming.
```
Comment (SunkissedQueen) [inline interview-questions.md:16]: Do you have a preference for either?

File: interview-questions.md [hunk 6]
```diff
@@ -8,42 +8,78 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+A float takes decimal values, however it needs to be declared when writing the method to print the decimal numbers. An example is: 
+p 9.0 / 5
+output ---> 1.8
+
+An integer, on the other hand, are whole numbers and can return negative or positive. Example:
+p 9 / 5
+output ---> 1
+
+p 2 - 10
+output ---> -8
```
Comment (SunkissedQueen) [inline interview-questions.md:29]: the examples make this very engaging

File: interview-questions.md [hunk 7]
```diff
@@ -8,42 +8,78 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 Researched answer:
+Based on what I read from resources online, float and integer are both numerical data. However, integers doesn't numbers with decimal points. It will just round it down. Float, on the other hand, return the numbers with the decimal points, giving an more detailed response.
```
Comment (SunkissedQueen) [inline interview-questions.md:32]: Use the research not only to confirm your understanding but to learn something new

File: interview-questions.md [hunk 8]
```diff
@@ -8,42 +8,78 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-Researched answer:
+def baby_name
+    'PJ'
+end
+p baby_name
+output --> "PJ"
+
+Researched answer: 
+Based on what I read from resources online, I found the same definition as I have that it gives an output without having to include the word "return". In addition to that definition, I also have read that Ruby automatically returns the value of the last line and writing the "return" is not necessary. Although it accepts the word "return", for best practice and keeping the code cleaner, the word "return" is not included.
```
Comment (SunkissedQueen) [inline interview-questions.md:44]: nice content

File: ruby-challenges.rb [hunk 9]
```diff
@@ -22,8 +33,15 @@
...
+def no_vowels (title)
+    title.delete('aeiouAEIOU')
+end
 
-
+p no_vowels(beatles_album1)
+p no_vowels(beatles_album2)
+p no_vowels(beatles_album3)
+p '-----no-Vowels-----'
+puts "\n"
```
Comment (SunkissedQueen) [inline ruby-challenges.rb:44]: I like that you are making use of the console

File: ruby-challenges.rb [hunk 10]
```diff
@@ -32,3 +50,27 @@
...
+            "#{word.capitalize} is not a palindrome word"
+        end
+    end   
+end
+
+p palindromeWord(palindrome_tester1)
+p palindromeWord(palindrome_tester2)
+p palindromeWord(palindrome_tester3)
+
+
```
Comment (SunkissedQueen) [inline ruby-challenges.rb:75]: Very nice. Definitely go back to refactor Glad you understood that all the letters needed to be either upper or lower case. Is there a way to either make all the letters upper or lower case and then use one if/else statement?

**Review by SunkissedQueen — APPROVED** [review]
Cathrine, this hard work is approved! You definitely understand how to manipulate the data. You have grasped the concept of the tools we shown you. Next week I would like to see you refactor using methods that will reduce your lines of code.  For the interview questions, I would say maintain your approach with showing examples with the definition. Since you are good with the vocabulary, please use the research to bring in interesting facts.

File: interview-questions.md (hunk 5 above)
Comment (CathrineC) [reply interview-questions.md:16]: Right now, I like OOP as it is the one that I am familiar with. However, I would definitely want to learn Functional programming.

**Pull request merged**